tracer/src/Datadog.Trace/ClrProfiler/AutoInstrumentation/Logging/NLog/DirectSubmission/Formatting/NLogLogFormatter.cs

```diff
@@ -0,0 +1,97 @@
+                GetLogLevelString(logEntry.Level),
+                logEntry.Exception,
+                RenderProperties);
+        }
+
+        private static LogPropertyRenderingDetails RenderProperties(JsonTextWriter writer, LogEntry logEntryWrapper)
+        {
+            var haveSource = false;
+            var haveService = false;
+            var haveHost = false;
+            var haveTags = false;
+            var haveEnv = false;
+            var haveVersion = false;
+
+            if (logEntryWrapper.Properties is not null)
+            {
+                foreach (var kvp in logEntryWrapper.Properties)
+                {
+                    var name = kvp.Key;
+                    haveSource |= LogFormatter.IsSourceProperty(name);
+                    haveService |= LogFormatter.IsServiceProperty(name);
+                    haveHost |= LogFormatter.IsHostProperty(name);
+                    haveTags |= LogFormatter.IsTagsProperty(name);
+                    haveEnv |= LogFormatter.IsEnvProperty(name);
+                    haveVersion |= LogFormatter.IsVersionProperty(name);
+
+                    LogFormatter.WritePropertyName(writer, name);
+                    LogFormatter.WriteValue(writer, kvp.Value);
+                }
+            }
+            else if (logEntryWrapper.FallbackProperties is not null)
+            {
+                foreach (var kvp in logEntryWrapper.FallbackProperties)
+                {
+                    var name = kvp.Key.ToString();
+                    haveSource |= LogFormatter.IsSourceProperty(name);
+                    haveService |= LogFormatter.IsServiceProperty(name);
+                    haveHost |= LogFormatter.IsHostProperty(name);
+                    haveTags |= LogFormatter.IsTagsProperty(name);
+                    haveEnv |= LogFormatter.IsEnvProperty(name);
+                    haveVersion |= LogFormatter.IsVersionProperty(name);
+
+                    LogFormatter.WritePropertyName(writer, name);
+                    LogFormatter.WriteValue(writer, kvp.Value);
+                }
+            }
+
+            return new LogPropertyRenderingDetails(haveSource, haveService, haveHost, haveTags, haveEnv, haveVersion, messageTemplate: logEntryWrapper.LogEventInfo.Message);
+        }
+
+        private static string GetLogLevelString(LogLevelProxy logLevel) =>
+            logLevel.Ordinal switch
+            {
+                0 => DirectSubmissionLogLevelExtensions.Verbose,
+                1 => DirectSubmissionLogLevelExtensions.Debug,
+                2 => DirectSubmissionLogLevelExtensions.Information,
+                3 => DirectSubmissionLogLevelExtensions.Warning,
+                4 => DirectSubmissionLogLevelExtensions.Error,
+                5 => DirectSubmissionLogLevelExtensions.Fatal,
+                // Technically there's a 6, off, but should never have this level in a log message
+                _ => throw new ArgumentOutOfRangeException(nameof(logLevel))
```

Comment:
No exceptions as discussed in other sections.

Reply:
Done in https://github.com/DataDog/dd-trace-dotnet/pull/1945/commits/7f655a451f63c85b0c2f961e65dd04892e996920